HLTrigger/HLTfilters/plugins/HLTDoubletDZ.cc

```diff
@@ -310,348 +342,271 @@ HLTDoubletDZ<T1, T2>::getCollections(edm::Event& iEvent,
-  std::vector<l1t::PFTauRef> coll2;
+  auto const& electrons = iEvent.get(electronToken_);
 
-  if (getCollections(iEvent, coll1, coll2, filterproduct)) {
-    int n(0);
-    l1t::PFTauRef r1;
-    l1t::PFTauRef r2;
+  bool hasValidVZ1 = false;
+  auto const vz1 = getCandidateVZ(c1, hasValidVZ1, minPixHitsForDZ_, electrons);
+  if (not hasValidVZ1)
+    return true;
 
-    for (unsigned int i1 = 0; i1 != coll1.size(); i1++) {
-      r1 = coll1[i1];
-      unsigned int I(0);
-      if (same_) {
-        I = i1 + 1;
-      }
-      for (unsigned int i2 = I; i2 != coll2.size(); i2++) {
-        r2 = coll2[i2];
+  bool hasValidVZ2 = false;
+  auto const vz2 = getCandidateVZ(c2, hasValidVZ2, minPixHitsForDZ_, electrons);
+  if (not hasValidVZ2)
+    return true;
 
-        if (!computeDZ(iEvent, r1, r2))
-          continue;
+  return (std::abs(vz1 - vz2) <= maxDZ_);
+}
 
-        n++;
-        filterproduct.addObject(triggerType1_, r1);
-        filterproduct.addObject(triggerType2_, r2);
-      }
-    }
+template <>
+bool HLTDoubletDZ<reco::RecoChargedCandidate, reco::RecoChargedCandidate>::computeDZ(
+    edm::Event const& iEvent, reco::RecoChargedCandidate const& c1, reco::RecoChargedCandidate const& c2) const {
+  if (not passCutMinDeltaR(c1, c2))
+    return false;
 
-    accept = accept || (n >= min_N_);
-  }
+  bool hasValidVZ1 = false;
+  auto const vz1 = getCandidateVZ(c1, hasValidVZ1, minPixHitsForDZ_);
+  if (not hasValidVZ1)
+    return true;
+
+  bool hasValidVZ2 = false;
+  auto const vz2 = getCandidateVZ(c2, hasValidVZ2, minPixHitsForDZ_);
+  if (not hasValidVZ2)
+    return true;
+
+  return (std::abs(vz1 - vz2) <= maxDZ_);
+}
 
-  return accept;
+template <>
+bool HLTDoubletDZ<l1t::TrackerMuon, l1t::TrackerMuon>::computeDZ(edm::Event const& iEvent,
+                                                                 l1t::TrackerMuon const& c1,
+                                                                 l1t::TrackerMuon const& c2) const {
+  return ((std::abs(c1.phZ0() - c2.phZ0()) <= maxDZ_) and passCutMinDeltaR(c1, c2));
 }
 
-/// Special instantiation for L1HPSPFTau
-/// L1HPSPFTau are *not* RecoCandidates, therefore they don't implement superCluster()
-/// They are LeafCandidates instead
 template <>
-bool HLTDoubletDZ<l1t::HPSPFTau, l1t::HPSPFTau>::hltFilter(edm::Event& iEvent,
-                                                           const edm::EventSetup& iSetup,
-                                                           trigger::TriggerFilterObjectWithRefs& filterproduct) const {
+bool HLTDoubletDZ<l1t::HPSPFTau, l1t::HPSPFTau>::computeDZ(edm::Event const& iEvent,
+                                                           l1t::HPSPFTau const& c1,
+                                                           l1t::HPSPFTau const& c2) const {
+  if (not passCutMinDeltaR(c1, c2))
+    return false;
+
+  bool hasValidVZ1 = false;
+  auto const vz1 = getCandidateVZ(c1, hasValidVZ1);
+  if (not hasValidVZ1)
+    return false;
```

Comment:
For `l1t::HPSPFTau`s, this PR makes the choice of rejecting the candidate if it cannot provide a valid dz measurement (b/c of invalid track ref). Such case would lead to a crash in the current implementation, since the Ref is not checked (and obviously Run-3 menus are unaffected by this choice).